### PR TITLE
feat: Oracle vote extension handler

### PR DESCRIPTION
**Pull request by rbajollari**
## Description

closes: #XXXX

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] added appropriate labels to the PR
- [ ] targeted the correct branch
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items._

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: buf.work.yaml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+version: v1
+directories:
+  - proto
```
Comment (adamewozniak) [inline buf.work.yaml:3]: Do we need this bit?

File: cmd/ojod/cmd/commands.go [hunk 2]
```diff
@@ -55,19 +56,31 @@ func initAppConfig() (string, interface{}) {
...
 		Config: *srvCfg,
 		WASM: WASMConfig{
 			LruSize:       1,
 			QueryGasLimit: 300000,
 		},
+		PriceFeeder: pricefeeder.AppConfig{
+			ConfigPath:     dir + "/ojo/pricefeeder/price-feeder.example.toml",
+			ChainConfig:    true,
+			LogLevel:       "info",
+			OracleTickTime: time.Second * 5,
```
Comment (adamewozniak) [inline cmd/ojod/cmd/commands.go:82]: Are these the default values?

File: config.yml [hunk 3]
```diff
@@ -33,7 +33,14 @@ genesis:
...
+    params:
+        abci:
+          vote_extensions_enable_height: "2"
   chain_id: ojo-testnet
 validators:
 - name: alice
   bonded: 33500000000000uojo
+  app:
+    pricefeeder:
+      config_path: "PATH TO PRICE FEEDER CONFIG HERE"
```
Comment (adamewozniak) [inline config.yml:46]: Can we set this to a default in the repo?
```suggestion
      config_path: "PATH TO PRICE FEEDER CONFIG HERE"
```

File: go.mod [hunk 4]
```diff
@@ -32,9 +33,10 @@ require (
 	github.com/golangci/golangci-lint v1.55.2
 	github.com/gorilla/mux v1.8.1
 	github.com/grpc-ecosystem/grpc-gateway v1.16.0
+	github.com/grpc-ecosystem/grpc-gateway/v2 v2.19.1
 	github.com/mgechev/revive v1.3.6
 	github.com/mitchellh/mapstructure v1.5.0
-	github.com/ojo-network/price-feeder v0.1.10-0.20240221234245-b3d7d675adc9
+	github.com/ojo-network/price-feeder v0.1.11-0.20240403153904-aeda5a5840a6
```
Comment (adamewozniak) [inline go.mod:39]: Before we merge this in we should tag it to a specific price feeder release (after [this gets merged](https://github.com/ojo-network/price-feeder/pull/370))

File: proto/buf.gen.pulsar.yaml [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+      - buf.build/googleapis/googleapis
+      - buf.build/cosmos/gogo-proto
+      - buf.build/cosmos/cosmos-proto
+    override:
+plugins:
+  - name: go-pulsar
+    out: ./api
+    opt: paths=source_relative
+  - name: go-grpc
+    out: ./api
```
Comment (adamewozniak) [inline proto/buf.gen.pulsar.yaml:21]:
```suggestion
```
Pretty sure we don't need this file, can probably gitignore

File: proto/buf.gen.sta.yaml [hunk 6]
```diff
@@ -0,0 +1,15 @@
+# This file is auto-generated from Ignite. You can edit
+# the file content but do not change the file name or path.
+#
+# buf.gen.sta.yaml
+#
+version: v1
```
Comment (adamewozniak) [inline proto/buf.gen.sta.yaml:6]: same here, probably don't need this or the other swagger files

**Review by adamewozniak — COMMENTED** [review]
Initial review

**Comment by adamewozniak** [comment]
Dependency issues will be resolved by merging https://github.com/ojo-network/ojo/pull/405/files

File: x/oracle/abci/abci_test.go [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+func MockPriceFeeder() *pricefeeder.PriceFeeder {
+	providers := make(map[types.ProviderName][]types.CurrencyPair)
+	deviations := make(map[string]math.LegacyDec)
+	providerEndpointsMap := make(map[types.ProviderName]provider.Endpoint)
+
+	oracle := oracle.New(
+		zerolog.Nop(),
+		client.OracleClient{},
+		providers,
+		time.Second*10,
```
Comment (adamewozniak) [inline x/oracle/abci/abci_test.go:47]: Why 10 seconds here instead of 5?

File: x/oracle/abci/abci_test.go [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+	suite.Suite
+
+	ctx  sdk.Context
+	app  *ojoapp.App
+	keys []integration.TestValidatorKey
+}
+
+func (s *IntegrationTestSuite) SetupTest() {
+	s.app, s.ctx, s.keys = integration.SetupAppWithContext(s.T())
+	s.app.OracleKeeper.SetVoteThreshold(s.ctx, math.LegacyMustNewDecFromStr("0.4"))
```
Comment (adamewozniak) [inline x/oracle/abci/abci_test.go:31]: can we set the threshold to a const in the testing suite

File: x/oracle/abci/preblocker.go [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+	"fmt"
+
+	"cosmossdk.io/log"
+	cometabci "github.com/cometbft/cometbft/abci/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/ojo-network/ojo/x/oracle/keeper"
+)
+
+type PreBlockHandler struct {
```
Comment (adamewozniak) [inline x/oracle/abci/preblocker.go:14]: Let's add a comment here explaining the object in detail

File: x/oracle/abci/preblocker.go [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+				h.logger.Error("failed to decode injected vote extension tx", "err", err)
+				return nil, err
+			}
+
+			// set oracle exchange rate votes using the passed in context, which will make
+			// these votes available in the current block.
+			for _, exchangeRateVote := range injectedVoteExtTx.ExchangeRateVotes {
+				valAddr, err := sdk.ValAddressFromBech32(exchangeRateVote.Voter)
+				if err != nil {
+					return nil, err
```
Comment (adamewozniak) [inline x/oracle/abci/preblocker.go:52]: If there is an invalid voter address to a validator address here, will it stop the entire preblock function? I think this should instead be skipped over and logged:

```suggestion
					h.logger.Error("failed to xyz", "err", err)
					continue
```

File: x/oracle/abci/preblocker.go [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+				}
+				h.keeper.SetAggregateExchangeRateVote(ctx, valAddr, exchangeRateVote)
+			}
+		}
+
+		h.logger.Info(
+			"oracle preblocker executed",
+			"vote_extensions_enabled", voteExtensionsEnabled,
```
Comment (adamewozniak) [inline x/oracle/abci/preblocker.go:60]: Am I right to assume that if this log happens, the "extensions enabled" boolean will always be true?
```suggestion
```

File: buf.work.yaml (hunk 1 above)
Comment (rbajollari) [reply buf.work.yaml:3]: For the new version of ignite that uses sdk v0.50 it requires these to be generated when I run ignite chain serve

File: cmd/ojod/cmd/commands.go (hunk 2 above)
Comment (rbajollari) [reply cmd/ojod/cmd/commands.go]: Took this out 👍

File: x/oracle/abci/preblocker.go (hunk 11 above)
Comment (rbajollari) [reply x/oracle/abci/preblocker.go:64]: This is outside of the if statement that checks vote extensions are enabled so it should still be executed if they're not

File: x/oracle/abci/abci_test.go (hunk 7 above)
Comment (rbajollari) [reply x/oracle/abci/abci_test.go]: Changed to 5

File: buf.work.yaml (hunk 1 above)
Comment (rbajollari) [reply buf.work.yaml]: Added to git ignore

File: config.yml (hunk 3 above)
Comment (rbajollari) [reply config.yml]: Updated to relative path

File: x/oracle/abci/preblocker.go (hunk 11 above)
Comment (adamewozniak) [reply x/oracle/abci/preblocker.go:64]: gotcha - ty

File: x/oracle/abci/proposal.go [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+			if err != nil {
+				return nil, err
+			}
+
+			injectedVoteExtTx := AggregateExchangeRateVotes{
+				ExchangeRateVotes:  exchangeRateVotes,
+				ExtendedCommitInfo: req.LocalLastCommit,
+			}
+
+			// TODO: Switch from stdlib JSON encoding to a more performant mechanism.
```
Comment (adamewozniak) [inline x/oracle/abci/proposal.go:84]: Let's add a reference to an open github issue here

File: x/oracle/abci/proposal.go [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+			bz, err := json.Marshal(injectedVoteExtTx)
+			if err != nil {
+				h.logger.Error("failed to encode injected vote extension tx", "err", err)
+				return nil, errors.New("failed to encode injected vote extension tx")
+			}
+
+			// Inject a "fake" tx into the proposal s.t. validators can decode, verify,
```
Comment (adamewozniak) [inline x/oracle/abci/proposal.go:91]:
```suggestion
			// Inject a placeholder tx into the proposal s.t. validators can decode, verify,
```

File: x/oracle/abci/proposal.go [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+
+	return votes, nil
+}
+
+func (h *ProposalHandler) verifyExchangeRateVotes(
+	injectedVotes []oracletypes.AggregateExchangeRateVote,
+	generatedVotes []oracletypes.AggregateExchangeRateVote,
+) error {
+	if len(injectedVotes) != len(generatedVotes) {
+		return errors.New("number of exchange rate votes in vote extension and extended commit info are not equal")
```
Comment (adamewozniak) [inline x/oracle/abci/proposal.go:223]: should we register this in errors.go?

File: x/oracle/abci/proposal.go [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+	}
+
+	for i := range injectedVotes {
+		injectedVote := injectedVotes[i]
+		generatedVote := generatedVotes[i]
+
+		if injectedVote.Voter != generatedVote.Voter || !injectedVote.ExchangeRates.Equal(generatedVote.ExchangeRates) {
+			return errors.New("injected exhange rate votes and generated exchange votes are not equal")
```
Comment (adamewozniak) [inline x/oracle/abci/proposal.go:231]: +1 same here

File: x/oracle/abci/voteextension.go [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+		if err != nil {
+			err := fmt.Errorf("extend vote handler received invalid exchange rate %w", types.ErrInvalidExchangeRate)
+			h.logger.Error(
+				"height", req.Height,
+				err.Error(),
+			)
+			return nil, err
+		}
+
+		// Filter out rates which aren't included in the AcceptList
```
Comment (adamewozniak) [inline x/oracle/abci/voteextension.go:85]:
```suggestion
		// Filter out rates which aren't included in the AcceptList.
```

File: x/oracle/types/params.go [hunk 17]
```diff
@@ -83,13 +83,24 @@ var (
 				"coinbase",
 			},
 		},
+		CurrencyPairProviders{
+			BaseDenom:  AtomSymbol,
+			QuoteDenom: USDDenom,
+			Providers: []string{
```
Comment (adamewozniak) [inline x/oracle/types/params.go:89]: Let's put the defaults as BTC, ETH, and ATOM here, with multiple providers plz

**Review by adamewozniak — APPROVED** [review]
Aside from one lint fix, lgtm

**Pull request merged**